### DISCUSSION TITLE
How to convert image to rgba8888 byte array?

### DISCUSSION THREAD

**Discussion by BenMcLean** [SixLabors/ImageSharp#1941] [Ideas] [answered]
I don't know where support questions go so I'm posting here ... let me know if there is a better place.

I've got SixLabors.ImageSharp 1.0.4 from NuGet.

In my program, the user submits a `.png` file.

I want to get the height, width and a byte array in the format of rgba8888, where every four bytes represents one pixel and only pixels are in the array (no header data, no footer data, no metadata, just pixels) so that if you divide the array size by 4 * height then you'd get width and if you divide the array size by 4 * width then you'd get height.

There are lots of answers on the Internet for how to do this, but they are all for old versions of ImageSharp and do not work on the current version.

Even [the documentation](https://docs.sixlabors.com/articles/imagesharp/pixelbuffers.html#exporting-raw-pixel-data-from-an-imaget) does not actually say directly how to do this. It has something about doing things one way with a normal image and another way with a multi-megapixel image and has you converting to a data type called `Rgba32` without specifying how to get a byte array from there.

I want a byte array because I want to write code that is agnostic to what image library is being used and will only include ImageSharp in my test project.

**Comment by antonfirsov** [comment, marked as answer]
I'm assuming that rgba8888 is equivalent of our `Rgba32` pixel type in my answer.

If your image is not insanely huge (above 130 Megapixels ~ 11000 x 11000 pixels), the `TryGetSinglePixelSpan` method should succeed and you should be able to cast the `Span<Rgba32>` to `Span<byte>`, and eventually copy it to a `byte[]` just like the docs suggest.

Note that these API-s are about to go in ImageSharp 2.0, because of the necessary changes of the underlying buffer management. There will [be `image.CopyPixelDataTo` methods](https://github.com/SixLabors/ImageSharp/blob/dfe3fb5387dc64a5e3849c3202d92f1c775d414c/src/ImageSharp/Image%7BTPixel%7D.cs#L304-L314) instead. See #1739 for more details.

> I want a byte array because I want to write code that is agnostic to what image library is being used and will only include ImageSharp in my test project.

We are intentionally lacking an `image.ToByteArray()` and similar API-s, because the large extra allocations hurt performance a lot as applications scale. If performance matters, I really not recommend to go with the `byte[]` approach. Instead, you can define your own abstraction layer with an image base class that defines *capabilities to enumerate pixel data*, though I admit this is not a trivial job.

**Reply by BenMcLean** [reply]
> We are intentionally lacking an image.ToByteArray() and similar API-s, because the large extra allocations hurt performance a lot as applications scale.

Based on this, I think ImageSharp might just flat out be the wrong library or tool for me to use altogether. It just will not do the job and I need to find something else that will.

**Reply by BenMcLean** [reply]
I just wanted to get the pixels out of a `.png` and it looks like doing that will require writing and maintaining my own library from scratch.

**Reply by antonfirsov** [reply]
Yes. I think it's possible to create a performant, yet relatively thin layer that can unify ImageSharp 1.0/2.0, SkiaSharp, System.Drawing etc.

Though if extra allocations are OK in your use case, you should be fine with a method that extracts `byte[]` from `Image<Rgba32>`. I recommend you to use [our MyGet feed](https://www.myget.org/feed/sixlabors/package/nuget/SixLabors.ImageSharp) to get the ImageSharp 2.0 API, so you don't have to update your code. The official NuGet release is just weeks away.

With our core API, we really prefer to only encourage techniques and practices that do not hurt scaling and performance, even if the lack of direct support for certain convenience techniques is a disadvantage for some users.

**Reply by antonfirsov** [reply]
> I just wanted to get the pixels out of a `.png` and it looks like doing that will require writing and maintaining my own library from scratch.

I don't see why is this the case.

**Reply by BenMcLean** [reply]
> Though if extra allocations are OK in your use case,

This is 1990s style retro pixel art for video games. ***Tiny*** images sizes. I think I'm good.

>  you should be fine with a method that extracts byte[] from Image<Rgba32>.

Assuming I got ImageSharp 2.0, what would that method look like?

Right now, when I've got a byte array of a picture and I want to see what it looks like, I just run this to see the output at the end of my program:

```csharp
SixLabors.ImageSharp.Image.LoadPixelData<SixLabors.ImageSharp.PixelFormats.Rgba32>(
  byteArray,//data
  width,//width
  height//height
  ).SaveAsPng("output.png");
```

That is literally the only job ImageSharp currently has and all I need it to do is the reverse operation of this.

**Reply by BenMcLean** [reply]
> I don't see why is this the case.

Sorry, I guess I'm having a bad day. I shouldn't have said "from scratch" since yeah, I could probably do the abstraction layer thing.

Am just feeling really irritated that I'm having so much trouble with a task that should be so simple, even if inefficient.

**Reply by antonfirsov** [reply]
> Assuming I got ImageSharp 2.0, what would that method look like?

Use the [`CopyPixelDataTo`](https://github.com/SixLabors/ImageSharp/blob/dfe3fb5387dc64a5e3849c3202d92f1c775d414c/src/ImageSharp/Image%7BTPixel%7D.cs#L314) method I mentioned. `T[]` is implicitly convertible to `Spans<T>`.

```C#
byte[] data = new byte[sizeof(Rgba32) * image.Width * image.Height];
image.CopyPixelDataTo(data);
```

**Reply by ghost** [reply]
You can export to bytes like this (although starting from version 2.0 of the ImageSharp, it is now not possible to do it
in parallel, and it is a huge drawback - i hope the authors will create a multythreaded version of ProcessPixelRows method "a pixelAccessor per thread" - now it is single threaded)

   ```
private static byte[] ToRGB24(Image<Rgb24> image)
        {
            int pixelSize = GetPixelSize(ColorDepths.Bits8, 3);
            int height = image.Height;
            int width = image.Width;
            int rowSize = width * pixelSize;
            byte[] result = new byte[rowSize * height];

            image.ProcessPixelRows(pixelAccessor =>
            {
                for (int y = 0; y < pixelAccessor.Height; y++)
                {
                    Span<Rgb24> rgbs = pixelAccessor.GetRowSpan(y);

                    Span<byte> resultSpan = result;
                    Span<byte> rowSpan = MemoryMarshal.AsBytes(rgbs);
                    Span<byte> outputSpan = resultSpan.Slice(y * rowSpan.Length, rowSpan.Length);
                    rowSpan.CopyTo(outputSpan);
                }
            });

            return result;
        }
```

**Reply by ghost** [reply]
P.S. - although  image.CopyPixelDataTo(result); looks like more efficient

**Comment by JimBobSquarePants** [comment]
Hey @BenMcLean do you have a Six Labors license? Because this all reads a lot like you’re asking for support.

**Reply by BenMcLean** [reply]
Yeah, I guess it would be a support request, but for non-commercial individual hobbyist open source development: not for a company. Didn't know about the licensing thing when I posted this. I guess that explains why there's not really much of a community forum or anything because the business model is to get people to buy the support licenses. Sorry, I want to respect that and don't mean to be a freeloader: was just trying to get a quick simple function out of the most obvious FOSS thing I found for it on Google.

**Reply by JimBobSquarePants** [reply]
No worries. https://github.com/SixLabors/ImageSharp/discussions/1941#discussioncomment-1958151 is the answer you need anyway.
